### ISSUE TITLE
Native tasks behave better when run in parallel

### ISSUE THREAD

**Issue by ghale** [gradle/gradle-native#70]
While working on #49, I found an issue with `DefaultBuildOperationQueue` where we actually perform worse when tasks execute in parallel.  The issue is that we create a separate queue for each task that runs in parallel.  Each queue submits up to max-workers Runnables to read from the queue.  However the queues use a shared fixed size executor pool that only has max-workers threads.  When running a single task at a time, this works fine, because all of the Runnables get executed and can acquire a worker lease.  However, when we have multiple tasks running in parallel, we get max-workers^2 Runnables trying to run in an executor pool with only max-workers threads.  The result is that some Runnables get executed but block trying to get a worker lease, while others (who could potentially borrow a worker lease from their parent) cannot execute because the thread pool is full.  Only one Runnable can borrow its parent's lease so it ends up processing all of a its queue's operations in a single thread.  Eventually, that queue finishes, it's Runnables exit and that unblocks another queue, etc etc, until everything finishes, but the whole build slows down because we don't make good use of the threads in the executor pool.

A couple of options:

- We have a fixed set of Runnables that know about all queues and process any work that is ready (i.e. we never have more than max-workers Runnables in the fixed size executor pool)
- We only start a Runnable in `DefaultBuildOperationQueue` if we can acquire a worker lease.  The downside to this is that if there is a delay in between operations arriving on the queue, we are holding the worker lease the whole time, preventing other work from running.
- We scale the Runnables according to our ability to acquire a worker lease.  With this we have a single thread per queue that checks to see if there is additional work available and if a worker lease can be acquired.  If so, it submits a Runnable that runs only until there is no more work in the queue and then exits.  If new work arrives (and a lease is available) a new Runnable gets started.  The downside to this is that it could be complex.
- (For completeness) We don't use a fixed size pool and let the worker leases control the parallelism.  The downside to this is that we could start up to max-workers^2 threads.

**Comment by adammurdoch** [comment]
Another option: have the thread blocked waiting for queued work to complete also do some of this work.

**Comment by ghale** [comment]
Right on, @adammurdoch.  Just doing a little spiking, having the thread that is blocked waiting for completion also take part in processing the work looks like the simplest solution for now.  I'm going to take that approach.

**Comment by ghale** [comment]
So, I've implemented having the thread blocked waiting for completion participate in doing the work (https://github.com/gradle/gradle/tree/gh/stable-native/better-queue), but the more I think about it, the more I think we need a general solution to the problem that could be used in multiple places.  

We have two concepts for limiting parallel execution that are competing:

- Executor thread pools - Given an unbounded queue of work items, this constrains how many threads can execute items concurrently.
- Worker leases - Given an unbounded set of threads, this constrains how many can be actively executing work at any point in time.

One way to look at it is that executor thread pools constrain work within a queue of work items, while worker leases constrain execution across thread pools.  Worker leases allow the sharing of leases across thread pools via child borrowing.  However, when the only way to acquire a lease is by borrowing (because they're all allocated), a bounded thread pool can be underutilized.  This is because that while we can guarantee a work item can borrow a worker lease at some point in time, there's no guarantee that every executing item can borrow a lease at the moment in time the executor decides to execute it.  Inevitably we'll execute a work item that will block and prevent another item (that would not block) from execution.  The solution above works around this by allowing the work to be executed in the thread pool the worker lease is being borrowed from (rather than the pool it would normally be executed in).  However, if we could simply make a better decision about when to execute a work item, we could ensure that we only execute an item that can obtain a worker lease at the time that we execute it and would not block.  

In fact, I think we have this same pattern in at least a couple of places.  We have the build operation queue (which handles native compilation, parallel dependency resolution, test report generation, etc), the worker API, and the task execution plan.  These all fit the pattern that they execute in a bounded thread pool, but require coordination with other pools to limit concurrent execution.  I think we can implement a general solution that could be used in all three places.

We need something like a `ConditionalExecutionQueue`.  This would be backed by a bounded thread pool (i.e. with a size of max-workers) and would accept submissions to a queue of `ConditionalExecution` objects.  These objects would define 3 things:

- A set of conditions that must be met before the item can be executed.
- An "action" to execute when those conditions have been met.
- A set of completion actions to take when execution has finished.

For instance, for the build operation queue and the worker api, these would translate to:

- The conditions are only that a worker lease be acquired (either by borrowing from the parent or by grabbing an available root lease).
- The action is the operation (or unit of work) itself.
- The completion action is to release the worker lease.

For the task execution plan this is more complex, but still fits the pattern:

- The conditions are that a worker lease can be acquired, that the project lock can be acquired, that the task's dependencies have executed, that it can execute concurrently with the other running tasks, and maybe a few more things that I'm not thinking of off the top of my head.
- The action is the execution of the task itself.
- The completion actions are to release the worker lease, release the project lock, mark the task as complete, enforce any finalizer tasks, etc.

The `ConditionalExecutionQueue` would simply start max-workers threads that iterate over the queue and look for work items whose conditions can all be met.  When they find one, they execute it followed by its completion actions.

This is, of course, exactly how the task execution plan works right now, except that it's all tied up in complexity specific to the task execution plan.  However, I think we could use some of the same concepts and infrastructure to implement this for the build operation queue and the worker api right now and then look at migrating the task execution plan over to this more general pattern later.

Note that the `BuildOperationQueue` would not be a `ConditionalExecutionQueue` itself, but just a grouping of items that are submitted to a shared `ConditionalExecutionQueue`.  Multiple `BuildOperationQueue`s would all submit their work items to the shared queue and then wait for just their items to complete.

Thoughts?

**Comment by ghale** [comment]
I've cut https://github.com/gradle/gradle/pull/2693 to work around this problem by having the thread waiting for work to complete also execute some of that work when none of the queues workers make it onto the executor thread pool.

**Comment by ghale** [comment]
This has been merged.

I also did a little more exploration last week on the topic above of creating a more general solution to the lease/thread pool problem.  I spiked an implementation of what I described above (basically an execution queue that checks each item for conditions to determine if they are ready for execution or not).  The first pass had each individual work item passed to the execution queue by itself.  This worked fine, except that we lost the batching capability we had with `BuildOperationQueue` and this creates a lot of additional contention on the resource lock coordinator (and slows down performance).  So instead of passing each item of work, I changed it to submit something that represents the queue of work that could be executed as a batch (i.e. if a worker lease can be acquired, then process items from the queue until the queue is empty).  This "queue of queues" approach works ok, but I had problems getting the synchronization right (and I'm still not convinced there aren't any deadlocks).  I wanted to timebox this work to avoid going too far down the rabbit hole, so I got something mostly working and then stopped.

Ultimately, I think the second approach is probably the most feasible.  One central issue of this infrastructure is the big fat lock in the resource lock coordinator and reducing the overall contention on it as much as possible.  We probably want to look at ways we could do more batching of executions in queues of work.  For instance, the worker API should probably move towards a queued approach that batches requests once it can acquire a worker lease.  We might also consider looking at ways the task execution queue might be batched while a project lock has been acquired (and can be retained).  Some good abstractions that generalize this behavior and make it easy to avoid deadlocks, etc, would probably be the first step.

**Comment by adammurdoch** [comment]
Some thought on this:

There are 3 steps to running a work action:

1. Analysis when the action is submitted, to determine whether its requirements have been satisfied, and bookkeeping to keep track of which requirements can and can't be satisfied at any given time. Requirements are things like "this thing is available while I'm running".
2. Running the action while its requirements are satisfied.
3. Analysis when the action completes, to determine which other actions can now proceed.

One requirement common to _all_ actions is that there is sufficient CPU and memory available to run the action, which we approximate by running no more than n actions at a given time.

However, this is the _only_ requirement for many of the actions, particularly the most fine grained and actions. So, an increment towards a more efficient solution would be to introduce a fast track for actions with no requirements (beyond that that be able to run), and a slow track for other actions.

When an action is submitted, if it has no requirements, it would be added to a single shared queue from which a single pool of worker threads are taking work. The worker threads would do roughly what they do now in the build queue implementation:

1. block until some actions are available on the queue.
2. take out a work lease.
3. run as many actions as possible.
4. release work lease and goto 1.

This is essentially a restructuring of the build queue implementation, so that we don't create a bunch of worker threads per queue and have them fight for work leases (and take up stack and heap space and thrash the CPUs and their caches and all that), and so that a worker thread can draw work from multiple queues and so we are more efficient for small build queues or when a particular build queue is nearly empty.

We'd still have the build queue abstraction for this first step, but just make it more efficient and prepare the way for further improvements. For actions with requirements, I would leave them alone for this step and just continue to channel them through the more complex infrastructure. This feels fine to me because these actions tend to be coarse grained and there are far fewer of them running at a given time and there are fewer of them in general.

To make these complex actions more efficient, I would look at reducing the contention during steps 1 and 3 of the action lifecycle. Right now, each worker thread fights with the other worker threads to perform step 1 and step 3. Instead, I would separate this out so that some other thread takes care of the analysis of step 1 and when the requirements of the action are satisfied, adds the action to the shared queue that we introduced in the previous change to make build queue more efficient. After completing the action, the worker thread would put the action on to a 'finished queue' and move on to the next action. Some other thread would take care of the bookkeeping of step 3 and unblocking whoever cares about the action.

An option here would be to have the submitting thread take care of step 1 and a waiting thread take care of step 3 (often these are the same thread). A worker thread that submits and waits for work would help run the actions it is waiting on instead, rather than give up its worker lease. With this step, there is still some contention at the coordination level but it never involves the worker leases and the same pool of worker threads do all the work.

We might then improve this further (if it's even enough of a problem) by introducing a single dedicated scheduling thread that takes care of step 1 and step 3 of the action lifecycle. A submitting thread would add a work action to either the shared work queue if the action has no requirements, or the scheduling queue if the action has requirements. The scheduling thread would block on this scheduling queue and make decisions about which actions to run when and assign requirements to each. When an action is ready, the scheduling thread adds the action to the shared work queue for execution. On completion of the action, the scheduling thread would take care of queuing up any actions that are now unblocked. A waiting thread that is not a worker thread would just wait for its stuff to complete, and a waiting thread that is a worker thread would run some of the work it is waiting for.

**Comment by ghale** [comment]
Thanks @adammurdoch!  Good stuff.  I'm going to create a separate issue for exploring this some more (I probably should not have mixed this in with this particular issue) and start a doc that we can use to collaborate on a design.  I don't think there's a burning need to solve this problem right away, so I think we can probably let it marinate a little bit and iterate on the design some before we decide to make some changes.
